app-experimental.tex

```diff
@@ -2395,3 +2395,219 @@ \section{Distributed Capabilities For Peripherals And Accelerators} % <<<
+Having this solution leads to an atomic change of CID and code capability -- a property we envision to be useful for secure compartmentalisation.
+
+The CID will substitute the otype bits (18 bits) in the 128-bit capability format.
+The otype bits are currently not well-used. The otype field indicates whether a capability is sealed and if so which “type” it has.
+There exists currently two fixed values that are used: one for unsealed capabilities and one for sealed entry (sentry) capabilities.
+There are 14 more reserved fixed values, which are not currently used. All other values are used as values for sealed capabilities.
+
+However, there is also the possibility to combine otypes and CIDs.
+One potential approach is to subdivide the 18 otype bits such that there is space for a few otypes and the remainder of the bits is dedicated to CIDs.
+This would allow sentries and CID sealing to be combined.
+
+In the following text, we propose to completely substitute otypes.
+All values but the zero CID (CID==0) are valid IDs.
+Furthermore, every CID sealed code capability already is a sealed entry capability.
+Please note that every capability within a compartment can be manipulated by the compartment itself.
+This also includes CID sealed code capabilities.
+In the CHERI world up to now, sentries cannot be manipulated.
+
+At the moment, this proposal is limited to the 128-bit capability format.
+However, the underlying mechanism to add CID bits to the meta information bits of the capability works for every format.
+In the 128-bit capability format, we propose to allocate 10 bits to the compartment ID, but we envision any number of bits fitting in the meta  infomration bits to a valid implementation of CID Sealing.
+This length is also known as the ACID\_LENGTH, and keep the remaining 8 bits reserved, which allows for 1024 CIDs to be encoded.
+These CIDs are referred to as the Architectural CIDs (ACIDs).
+Software may choose to virtualise these and can create Software CIDs (SCIDs), which are a concept similar to PIDs.
+A SCID may consist of the corresponding ACID added to other bits in order to create a virtual identifier or can be fully independent of the ACID is mapped to.
+
+Compartment IDs come with two new instructions.
+One instruction for reading the CID into a general purpose register and one instruction for setting the CID of a capability.
+Like capabilities themselves, CIDs are not considered secret.
+Thereforem, the CID reading instruction is not privileged, but the writing instruction is restricted (see below).
+We suggest encodings for the instructions in order to demonstrate the adaptability of CID to the currently existing CHERI-RISC-V ISA.
+
+\texttt{CSetCID cd, rs0}:\\
+set the CID in cd to the value in rs0.
+This instruction needs the PERMIT\_SET\_CID bit set, otherwise it throws an exception.
+This instruction uses the ACID\_LENGTH lower bits of rs0 and ignores the upper bits.
+
```

Comment:
Do you need an authorising capability source operand where the PERMIT\_SET\_CID bit lives? This also suggests anybody with a capability with PERMIT\_SET\_CID can set any CID on a destination capability. Do you also mean to have an authorised CID range in the authority operand?

Reply:
I see further down that PERMIT_SET_CID lives in PCC. I think this should be made explicit when PERMIT_SET_CID is first mentioned (here). My question about the CID range remains for now...

Reply:
I have been pondering about this and I think we need to discuss this in person.